Add Framing Tool

digitalcarp opened the pull request:
# Framing Tool

Adds a framing/borders tool under the resize tool. Lightroom and darktable have similar functionality.

Plenty of people share images on platforms like Instagram. The platforms restrict image sizes and aspect ratios. This features allows you to keep your image's aspect ratio by adding a simple colored border. Users can have the tool downscale their images to fit within certain dimensions. This way the user can use RawTherapee's resizing method and post-resize sharpening tools rather than relying on the platform's algorithms.

The standard framing method can be used to add buffer areas for full-bleed printing (#6791).

It's easier to do batch processing for regular users when you don't need to use Gimp scripting or ImageMagick to add borders on multiple images.

## UI

![tool_default](https://github.com/user-attachments/assets/e9dc20b1-6914-45a2-aa9d-d1e1e92b97ec)

![tool_absolute](https://github.com/user-attachments/assets/175ea730-1c24-4395-8e12-c5121fa9850c)

![tool_bbox](https://github.com/user-attachments/assets/52a725b5-27eb-4c31-9ccd-1e157bc67345)

![tool_fixed_frame](https://github.com/user-attachments/assets/dcd441ac-700a-4c39-b2ab-1cb5b8f2b489)

## Processing Pipeline Changes

Framing resize operation does not perform resizing on an already resized image. Instead, if needed, the initial resize parameters are adjusted before performing the actual resize.

Post-resize sharpening will apply to the framing resized image. The border will not be in the image data yet, so sharpening will not be affected by the framing tool.

The borders are drawn after the image is in the output colorspace (i.e. after `lab2RgbOut()`).

## User Documentation

The framing tool adds a variable-sized border to your output image. It works in tandem with the Resize tool to downscale/upscale your image. Post-resize sharpening may be applied to the image as well.

### Tool Settings

#### Method

* Standard - Add a border around the image. The image's size is preserved when adding the border.
* Fixed Frame - Fills a fixed-size frame with the border and places the image in the center of the frame. The image may be shrunk to fit the fixed frame dimensions and any border sizing requirements.
* Bounding Box - Add a border around the image. If the framed dimensions including image and borders exceeds the bounding box, the image (measured including borders) will be shrunk to fit the bounding box.

#### Aspect Ratio

Specify the aspect ratio of the frame in Standard and Bounding Box framing methods. If the border sizing constraint is too small to meet the aspect ratio (i.e image would be cut off), then the minimum borders needed to meet the aspect ratio is used. The aspect ratio of the post-Resize tool image can be used by selecting the "As Image" option.

#### Orientation

Specify the orientation of the frame aspect ratio. The orientation of the post-Resize tool image can be used by selecting the "As Image" option.

#### Framed Width/Height

Specify the bounding box of the frame in pixels for the Bounding Box framing method. Specify the fixed frame dimensions in pixels for the Fixed Frame method.

#### Allow Upscaling to Frame

In the Fixed Frame and Bounding Box framing methods, if the image is smaller than the maximum allowed size, upscale the image to its max size.

#### Sizing Method

Determines how the size of the border is calculated. The border sizing values refer to the size on one side of the image. For example, a 10px border will add 20px the the total height of your image after framing.

##### Relative

Specify the border size based on a percentage of the chosen.

###### Basis

* Auto - Choose whichever side is the limiting side for small border sizes based on aspect ratios.
* Width/Height - Choose the width/height as the reference length.
* Long/Short Edge - Choose the longer/shorter edge as the reference length.

If the basis side border is too small for the requested aspect ratio, the minimum possible value will be used.

###### Limit Minimum Size

If enabled, ensures a minimum border size for horizontal/vertical borders as specified by the Minimum Width/Height inputs.

##### Uniform Relative

Adds a uniform border that is the same size horizontally and vertically. The border size is determined based on the image size as in the Relative sizing method. The minimum width and height are automatically updated to match each other.

##### Absolute

Specify the fixed border size in pixels using the Border Width/Height inputs. The aspect ratio and orientation options are ignored in this mode.

#### Border Color

Choose the 16-bit RGB values of the border.

### Limitations

* Downscaling limit is 32x32px (as in resize tool)
  * If the crop size is under 32x32, no downscaling is performed even if needed to fit frame.
  * Usually hit if the border size requirements are too high
* Upscaling limit is 16x crop size (as in resize tool)

## Example Output

Framing Method: Standard
Sizing: Relative @ Auto 10%
Crop Dimensions: 4800x3840
Resize Dimensions: 4800x3840
Adjusted Dimensions: 4800x3840
Output Dimensions: 5760x4608
![DSC_0066](https://github.com/user-attachments/assets/5b56f376-bbb3-48da-b3dd-2b19f6f3a527)

Framing Method: Standard
Sizing: Relative @ Auto 10%
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 500x400
Output Dimensions: 600x480
![DSC_0066-1](https://github.com/user-attachments/assets/a2ad8113-a083-41cd-9d88-f018c8d6eb78)

Framing Method: Standard (4:5 Portrait)
Sizing: Relative @ Height 10%
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 500x400
Output Dimensions: 500x625
![DSC_0066-2](https://github.com/user-attachments/assets/b7763a23-578d-479c-87ad-e5e1c0498b9e)

Framing Method: Standard (4:5 Portrait)
Sizing: Relative @ Height 10%
Min Size: W=200 H=100
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 500x400
Output Dimensions: 900x1125
![DSC_0066-3](https://github.com/user-attachments/assets/04ea50f9-8d3c-4789-a408-348f34295fde)

Framing Method: Standard
Sizing: Absolute @ W=100 H=200
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 500x400
Output Dimensions: 700x800
![DSC_0066-4](https://github.com/user-attachments/assets/b74b4c78-5008-4e13-8547-f040ae776c9a)

Framing Method: Fixed Frame (400x500)
Sizing: Relative @ Auto 10%
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 333x267
Output Dimensions: 400x500
![DSC_0066-5](https://github.com/user-attachments/assets/ada3c06e-8520-41d1-bfdc-24e40346d211)

Framing Method: Bounding Box (500x400) (4:5 Portrait)
Sizing: Relative @ Auto 10%
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 267x213
Output Dimensions: 320x400
![DSC_0066-6](https://github.com/user-attachments/assets/68dc1fdf-06ef-44b4-83d9-bd9df7ba035d)

Framing Method: Bounding Box (800x1000) (4:5 Landscape)
Sizing: Relative @ Auto 10%
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 500x400
Output Dimensions: 600x480
![DSC_0066-7](https://github.com/user-attachments/assets/42c49344-d5d0-4192-8dcb-ed49639b00c1)

Framing Method: Bounding Box (800x1000) (4:5 Landscape)
Sizing: Relative @ Auto 10%
Allow Upscaling To Frame: Yes
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 667x533
Output Dimensions: 800x640
![DSC_0066-8](https://github.com/user-attachments/assets/cb06e6e8-f692-41c4-a8bb-f82632192e85)

Framing Method: Fixed Frame (80x100)
Sizing: Absolute @ W=100 H=200
Crop Dimensions: 4800x3840
Resize Dimensions: 500x400
Adjusted Dimensions: 40x32
Output Dimensions: 80x100
![DSC_0066-9](https://github.com/user-attachments/assets/3d8e8364-2e50-4939-87af-964d0151e0ad)

Lawrence37 commented:
Thank you very much for working on this!

I briefly tried the framing feature. I have some feedback for you to consider while I continue testing.

1. The resize is applied twice while the image area and frame size are correct. For example, with a starting image of size 1000x1000 pixels, a scale factor of 0.50, and an absolute border size of 100 pixels, the resulting image has a total size of 700x700 pixels. The perimeter consists of a 100 pixel wide border and the interior image area is 500x500 pixels. This part is correct. However, the photo is 250x250 pixels in the upper-left corner of the image area. The rest is filled with black.
2. A border/frame tool is a frequently-requested feature. In the current state, the frame tool is not easily discoverable because it is located within the resize tool. This can be improved by either making the framing tool stand-alone in the GUI or by renaming the resize tool to something like "Resize & Framing".
3. In the absence of a way to see the border in the editor preview, it is hard to visualize the border color. Implementing the editor preview is complex, so I don't expect it to happen soon. A color indicator in the framing tool could be enough for now.
4. Some, and maybe most, users will find it strange that the color values range from 0 to 65535. Perhaps the GUI should present it as a percentage (i.e [0-100]) or with the more familiar [0-255] range?
5. The documentation should say that the border color values are in linear space with the primaries of the output profile (at least, that is what I observed). I welcome any discussion about what color space to use. The primaries of the output profile makes sense to me. I'm not so sure about using linear space instead of the output profile's TRC.
6. It would be nice if there was an option to apply a uniform border relative to the image size. For example, an image that is 4000x3000 pixels with a border that is 0.05 times the long edge would result in a border that is 200 pixels thick on all sides. This would be great for default processing profiles because the cropped and scaled image dimensions are not known in advanced and one might want a consistent look.

Lawrence37 commented:
I shared this pull request in the forum because I think it deserves exposure (no pun intended). https://discuss.pixls.us/t/border-framing-tool-in-development/46636

digitalcarp commented:
1. Fixed
2. Renamed to "Resize & Framing" as framing interacts with resize and post-resize sharpening
3. Added a color indicator on the bottom
5. Changed to [0, 255] range
5. I'm not an expert on color theory so you'll have to provide guidance for any changes you'd like. I'm happy to leave it as is and adjust in a future PR if required.
6. Implemented as "Uniform Relative". Updating PR user doc section to reflect.
7. Fixed


Examples for 6)

Framing Method: Standard
Sizing: Relative @ Auto 10%
Crop Dimensions: 1250x1000
Resize Dimensions: 625x500
Adjusted Dimensions: 625x500
Output Dimensions: 750x600
![DSC_0066-10](https://github.com/user-attachments/assets/dee57d93-a334-4647-8a61-7951727ae4f5)

Framing Method: Standard
Sizing: Uniform Relative @ Auto 10%
Crop Dimensions: 1250x1000
Resize Dimensions: 625x500
Adjusted Dimensions: 625x500
Output Dimensions: 750x625
![DSC_0066-11](https://github.com/user-attachments/assets/0bf5e4fa-fa61-4595-a7bf-7a2ce28b3475)

Lawrence37 commented:
Nice! Let me review.

Lawrence37 commented:
The fixes and improvements are working as expected. There are a few more issues I found, listed here with the associated test case:

1. With a 3:2 image, standard method, 4:3 aspect ratio (as image), relative sizing to the short edge, 0.05 size: There is no border.
2. With standard method, relative sizing, minimum width/height of 1 or 2: The minimum is not enforced.
3. With standard method, uniform relative sizing: The basis has no effect. The size is always relative to the image width.
4. With a 3:2 image, 200x200 bounding box, uniform relative sizing to the short edge, 0.05 size: There is no left/right border even if a minimum width is set.
5. With a 3:2 image, 200x200 bounding box, uniform relative sizing to the short edge, 0.5 size: The total image size is 96x80, which is too small.
6. With a 3:2 image, 200x200 fixed frame, uniform relative sizing, auto basis, 0.05 size: There is no left/right border. With a 2:3 image, there is no top/bottom border.

Regarding the color, the indicator works fine if the output and display profiles are sRGB. It will need to be more sophisticated to handle other profiles. I can help with that.

Lawrence37 commented:
I confirm 1, 2, 4, 5, and 6 are fixed. With the standard sizing and uniform relative sizing, the basis still has no effect. This time, the border width is always relative to the long edge.

Lawrence37 commented on rtengine/ipresize.cc:
```diff
@@ -458,4 +1098,93 @@ void ImProcFunctions::resize (Imagefloat* src, Imagefloat* dst, float dScale)
+
+    // Color::gamma2curve expects a 16-bit value, but the GUI sliders are
+    // using 8-bit values. Step up the user value to 16-bits.
+    auto clip = [](int v) { return std::max(0, std::min(v, 255)) * 256; };
```
Multiplying by 256 means it's not possible to have a white border in a 16-bit image since white is 65535.

digitalcarp replied:
Fixed with something along the lines of `return (v / 255) * 65535)`

Lawrence37 commented on rtengine/ipresize.cc:
```diff
@@ -458,4 +1098,93 @@ void ImProcFunctions::resize (Imagefloat* src, Imagefloat* dst, float dScale)
+        return rgb;
+    }
+
+    Imagefloat* framed = new Imagefloat(dims.framedWidth, dims.framedHeight);
```
Possible memory leak of `rgb`?

digitalcarp replied:
Fixed

Lawrence37 commented on rtengine/procevents.h:
```diff
@@ -1177,6 +1177,24 @@ enum ProcEventCode {
+    EvFramingAbsHeight = 1164,
+    EvFramingBorderRed = 1165,
+    EvFramingBorderGreen = 1166,
+    EvFramingBorderBlue = 1167,
```
Adding events here is error-prone. The new and recommended way is to create the events like how it's done [here](https://github.com/Beep6581/RawTherapee/commit/c166e0a7ed647b1d18a41f3cfb87a152d7ac36b1#diff-2568a327a7de95f76b48d0b558617b4c85478d7ee6403e1be2f64d13625c19afR30).

digitalcarp replied:
Fixed

Lawrence37 commented on rtengine/simpleprocess.cc:
```diff
@@ -1893,20 +1893,50 @@ class ImageProcessor
+        printf("  Framed image size: w=%d h=%d s=%f\n",
+               framingData.imgWidth, framingData.imgHeight, framingData.scale);
+        printf("  Total size: w=%d h=%d\n",
+               framingData.framedWidth, framingData.framedHeight);
```
Make sure these messages only print when verbose mode is enabled.

digitalcarp replied:
Fixed

digitalcarp commented:
> I confirm 1, 2, 4, 5, and 6 are fixed. With the standard sizing and uniform relative sizing, the basis still has no effect. This time, the border width is always relative to the long edge.

Yeah I observed that plus other sizing issues with uniform relative mode. I reworked the computations to fix a variety of scenarios (see framing-tests test21 to test-27).

Lawrence37 commented:
Looks like it works now. The code looks good to me. I'm going to retest to make sure everything else is still working as expected.

Lawrence37 commented:
I found one more issue. Using a 2:3 image, 200x200 bounding box, 1:1 ratio, relative sizing to the width, and 0.05 size, the minimum border size doesn't work with small values (like 2 pixels).
[bounding-box-minimum-size.pp3.txt](https://github.com/user-attachments/files/18003630/bounding-box-minimum-size.pp3.txt)

digitalcarp commented:
Fixed (see tests 28/29 in framing-tests)

Lawrence37 commented:
Looks good to me!

Lawrence37 commented:
Feel free to add yourself to AUTHORS.txt.

ror6ax commented:
Love this!!! Thank you!
